src/main/java/dev/ayu/latte/kafka/KafkaConnection.kt

```diff
@@ -0,0 +1,250 @@
+    val currentClusterId: Int,
+) {
+
+    private var producer: KafkaProducer<String, String>? = null
+    private var consumer: KafkaStreams? = null
+    private var shutdownHook: Thread? = null
+
+    private val subscribedTopics = Collections.synchronizedSet(mutableSetOf<String>())
+
+    private val topicListeners = Collections.synchronizedMap(HashMap<String, MutableSet<TopicListener>>())
+
+    private val eventDispatcherScope = CoroutineScope(Dispatchers.Default + SupervisorJob())
+    private val eventDispatcherErrorHandler = CoroutineExceptionHandler { _, t ->
+        log.error("Uncaught error in internal Kafka record handler", t)
+    }
+
+    private val isRunning: Boolean
+        // The `shutdownHook` is assigned a value as the last step of `start()`
+        get() = shutdownHook != null
+
+    fun send(
+        topic: String,
+        key: String,
+        value: String,
+        headers: KafkaRecordHeaders,
+    ): String {
+        val producer = this.producer ?: throw IllegalStateException("Producer is not initialized")
+        val record = ProducerRecord(topic, key, value)
+        headers.applyTo(record.headers())
+        producer.send(record).get()
+        // If the record is meant for ourselves (amongst other clusters),
+        // immediately dispatch it to the listeners.
+        if (headers.targetClusters.isEmpty() || currentClusterId in headers.targetClusters) {
+            invokeCallbacks(topic, key, value, headers)
+        }
+        return headers.requestId
+    }
+
+    fun on(topic: String, cb: TopicListener) {
+        val listeners = topicListeners.computeIfAbsent(topic) {
+            if (isRunning) {
+                throw IllegalStateException("Cannot subscribe to new topic after KafkaConnection has started")
+            }
+            subscribedTopics.add(topic)
+            CopyOnWriteArraySet()
+        }
+        listeners.add(cb)
+    }
+
+    fun off(topic: String, cb: TopicListener) {
+        topicListeners.computeIfPresent(topic) { _, listeners ->
+            listeners.remove(cb)
+            if (listeners.size == 0) {
+                subscribedTopics.remove(topic)
+                null
+            } else {
+                listeners
+            }
+        }
+    }
+
+    fun start() {
+        check(!isRunning) { "KafkaConnection is already running!" }
+        log.debug("Starting Kakfka Connection")
+        createTopics()
+        createProducer()
+        createConsumer()
+
+        shutdownHook = Thread({
+            shutdownHook = null
+            destroy()
+        }, "Kafka Connection ($clientId) Shutdown Hook")
+        Runtime.getRuntime().addShutdownHook(shutdownHook)
+        log.debug("Kakfka Connection is fully initialized")
+    }
+
+    fun destroy() {
+        consumer?.close()
+        consumer = null
+        producer?.close()
+        producer = null
+        shutdownHook?.let { Runtime.getRuntime().removeShutdownHook(it) }
+        shutdownHook = null
```

Comment:
If you mean like
```kt
shutdownHook?.let {
    // ...
    it = null
}
```
That doesn't work because `it` is a function argument (of the anonymous callback function), and function areguments are `val`, i.e. constants and can't be reassigned. Even if it were `var`, it would only change the local variable and not the original `shutdownHook`.